### DISCUSSION TITLE
Is this correct ?

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#478] [Q&A]
1) In Epoch description, in the Species paragraph we find an example  :
```
begin:species
   name = Electron
   charge = -1.0
   mass = 1.0
   frac = 0.5
   # npart = 2000 * 100
   number_density = 1.e4
   temp = 1e6

   temp_x = 0.0
   temp_y = temp_x(Electron)
   number_density_min = 0.1 * den_max
   number_density = if(abs(x) lt thick, den_max, 0.0)
   number_density = if((x gt -thick) and (abs(y) gt 2e-6), \
                        0.0, number_density(Carbon))
end:species

begin:species
   name = Carbon
   charge = 4.0
   mass = 1836.0*12
   frac = 0.5

   number_density = 0.25*number_density(Electron)
   temp_x = temp_x(Electron)
   temp_y = temp_x(Electron)

   dumpmask = full
end:species
```
There density of electrons defined via density of Carbon which is defined after that. But density of Carbon is defined via density of electron. Am i losing something?

2) Is there IF function without ELSE ? Why we always have to worry about ELSE ? Functions like this : 
    number_density = if(abs(x) lt thick, den_max)

**Comment by Status-Mirror** [comment]
This is likely a typo in our documentation, which I have now corrected. The previous input deck would probably run, but the behaviour would be unpredictable. Good spot!

The EPOCH maths parser only supports the `if(condition, value if true, value if false)` format. Extensions to the maths parser are difficult, and there's nothing you can't do with the current if statements that you could do with the proposed short one. If this is something you _really_ want, then we offer documentation on extending the maths parser here: https://epochpic.github.io/developer/extensions/maths_parser.html

**Comment by DanRRRR** [comment]
Thanks. Extending math parser could be even more the step into wrong direction. My point is that current IF construct complicates the My experience says that probably 20-30% of CPU time with my different codes is lost due to errors in setup. By the away Is IF FALSE part optional ?

EPOCH probably has setup difficulty 8 out of 10. Not in the sense it is so dark and complex but that it is easy to make a move you did not intended and did not do what you intended to do. At least initially when you learn how to use it. All is kind of acceptable, though it is pity to lose a day and find the error next day. But in some cases it becomes a disaster when you got just one day in a year on 1 million cores machine and found next day that wasted it due to setup error. I am just learning and i already lost on this IF-True_False couple weeks. I used two IFs and the next one's False part  interfered with the previous one ;( . It is the same like to give a small child complex task: go there and  if you find this - then do this and if that then this. 50% chance is he will fail. We all are always like that small children every time with the new confusing task  

As to Epoch unjustified complexities, Epoch has initial data and output in one directory, EXE is in other and it starting from the third, all with specific names. If i want all to be on one arbitrary directory - what i need to do, ideally with just one click ?

**Reply by Status-Mirror** [reply]
I believe it would be more confusing to current users if we suddenly changed the definition of maths parser elements. We have no plans to change the IF definition. I would always advise users to test their input decks before running a large simulation - especially on a 1 million core machine!

Our documentation tells users to run from a directory such that the code can be executed using `mpirun ./bin/epoch2d` (or 1d or 3d), as this is the simplest way to run the code. By doing this, the code is able to find the physics lookup tables for qed emission, bremsstrahlung, and ionisation. If you choose to not run from this directory, you will have to tell EPOCH where the lookup tables are using a command like: `physics_table_location` in the control block. Some specific packages also require an additional location flag.

**Comment by DanRRRR** [comment]
I wrote before about IF construct of math parser (see #412 and after 2 years of using  parser even more convinced that it was huge design defect to offer for it only one option 

if(A) Then B Otherwise C

This "Otherwise C" should be optional like all languages allow. For example in Fortran you can write either
1) IF(A) then B
or if you need more then  you can use
2) IF(A) then B  Else C 
which is full analog of EPOCH parser

So my suggestion for developers would be to modify parser so that "Otherwise C" would be optional

Because this "Otherwise C" makes setup a hell when multiple species are present. You scratch the head for hours and do multiple tries losing days because in supercomputers often see results only next day. You always have to figure out what to put into this "otherwise C" when you have 3-5 different species. And the solution looks totally unnatural and redundant. And even feels like being dumb and wrong 

Here is an example: suppose you need to setup 4-layers "cake" of different element
if "Otherwise C"  was optional (or would not exist at all because it is actually NOT needed) you just would set
for specie1
density = IF( (x gt x1) and (x lt x2) , density1 )
For specie2 
density = IF( (x gt x2) and (x lt x3) , density2 )
For specie3 
density = IF( (x gt x3) and (x lt x4) , density3 )
For specie4 
density = IF( (x gt x4) and (x lt x5) , density4 )

That's all you need ! But with current method though

For specie1
density = IF( (x gt x1) and (x lt x2) , density1, OTHERWISE WHAT? Why i was asked to do something here for other species ???? So many of them)
For specie2 
density = IF( (x gt x2) and (x lt x3) , density2, OTHERWISE WHAT the heck to put here - so many other species??)
For specie3 
density = IF( (x gt x3) and (x lt x4) , density3,  OTHERWISE WHAT put here??? )
For specie4 
density = IF( (x gt x4) and (x lt x5) , density4,  OTHERWISE WHAT ????? )

EPOCH is great and fast. Really worst part of EPOCH is this dumb IF

**Reply by Status-Mirror** [reply]
Hey Dan,

I personally don't find your example very convincing. When EPOCH loads particles into a species, it needs to know how many particles to load in EVERY cell.

Suppose we look at your line: `density = IF( (x gt x1) and (x lt x2) , density1 )`. When EPOCH comes to look at a cell with $x < x_1$ or $x > x_2$, then how can EPOCH know what density to use - you haven't specified anything. 

From the phrasing of your argument, I would assume you expect there to be no particles outside this range. In which case, a simple correction `density = IF( (x gt x1) and (x lt x2) , density1, 0 )` produces valid EPOCH syntax. This answers your "Otherwise what" question.

Cheers,
Stuart

P.S. I think I may have missed your confusion. When you specify density, you're only giving the density for a single particle species - not every particle species simultaneously. Hence, it doesn't matter how many other particle species are present.
